packages/muelu/src/Graph/MatrixTransformation/MueLu_DroppingCommon.hpp

```diff
@@ -0,0 +1,258 @@
+      auto clid = row.colidx(k);
+      if (rlid == clid) {
+        results(offset + k) = KEEP;
+        break;
+      }
+    }
+    return false;
+  }
+};
+
+template <class local_matrix_type>
+class MarkSingletonFunctor {
+ private:
+  using scalar_type        = typename local_matrix_type::value_type;
+  using local_ordinal_type = typename local_matrix_type::ordinal_type;
+  using memory_space       = typename local_matrix_type::memory_space;
+  using results_view       = Kokkos::View<DecisionType*, memory_space>;
+
+  using boundary_nodes_view = Kokkos::View<bool*, memory_space>;
+
+  local_matrix_type A;
+  boundary_nodes_view boundaryNodes;
+  results_view results;
+
+ public:
+  MarkSingletonFunctor(local_matrix_type& A_, boundary_nodes_view boundaryNodes_, results_view& results_)
+    : A(A_)
+    , boundaryNodes(boundaryNodes_)
+    , results(results_) {}
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  bool operator()(local_ordinal_type rlid) const {
+    auto row            = A.rowConst(rlid);
+    const size_t offset = A.graph.row_map(rlid);
+    for (local_ordinal_type k = 0; k < row.length; ++k) {
+      auto clid = row.colidx(k);
+      if ((results(offset + k) == KEEP) && (rlid != clid))
+        return false;
+    }
+    boundaryNodes(rlid) = true;
+    for (local_ordinal_type k = 0; k < row.length; ++k) {
+      auto clid = row.colidx(k);
+      if (rlid == clid)
+        results(offset + k) = KEEP;
+      else
+        results(offset + k) = BOUNDARY;
+    }
+    return true;
+  }
+};
+
+}  // namespace Misc
+
+namespace MatrixConstruction {
+
+template <class local_matrix_type, class FunctorsType>
+class PointwiseCountingFunctor {
+ private:
+  using scalar_type        = typename local_matrix_type::value_type;
+  using local_ordinal_type = typename local_matrix_type::ordinal_type;
+  using memory_space       = typename local_matrix_type::memory_space;
+  using results_view       = Kokkos::View<DecisionType*, memory_space>;
+
+  using rowptr_type = typename local_matrix_type::row_map_type::non_const_type;
+
+  local_matrix_type A;
+  results_view results;
+  rowptr_type rowptr;
+  FunctorsType functors;
+
+ public:
+  PointwiseCountingFunctor(local_matrix_type& A_, results_view& results_, rowptr_type& rowptr_, FunctorsType& functors_)
+    : A(A_)
+    , results(results_)
+    , rowptr(rowptr_)
+    , functors(functors_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const local_ordinal_type rlid, local_ordinal_type& nnz, const bool& final) const {
+    bool doneWithRow = false;
+    if constexpr (std::tuple_size<FunctorsType>::value >= 1) {
+      auto functor = std::get<0>(functors);
+      doneWithRow  = functor(rlid);
+    }
+    if constexpr (std::tuple_size<FunctorsType>::value >= 2) {
+      if (!doneWithRow) {
+        auto functor = std::get<1>(functors);
+        doneWithRow  = functor(rlid);
+      }
+    }
+    if constexpr (std::tuple_size<FunctorsType>::value >= 3) {
+      if (doneWithRow) {
```

Comment:
this would be a nasty bug to try to figure out
```suggestion
      if (!doneWithRow) {
```